packages/core/src/modules/connections/repository/ConnectionRecord.ts

```diff
@@ -20,8 +19,8 @@ export interface ConnectionRecordProps {
   theirDidDoc?: DidDoc
   theirLabel?: string
   invitation?: ConnectionInvitationMessage
-  state: ConnectionState
-  role: ConnectionRole
+  state: ConnectionState | DidExchangeState
+  role: ConnectionRole | DidExchangeRole
```

Comment:
Hmm. That does not seem so good. These are part of the protocol and protocol defines values of them. If I'm using older `connections` protocol I would expect the same values as they're defined in the spec. What about something like protocol metadata? 

I realize that in this context, we don't assume the older protocol to be there forever, it's hard to say how long it will be, hard to predict from my perspecitve.

Reply:
I tend to disagree. As a framework user I want the API to be as simple as possible. Having two state values that basically mean the same but have different values is not helping towards making a simple API. I think having a single state (our framework state) that is the default that can map to the 0160 state would be more helpful here. Especially having a single value to store will reduce complexity around the framework (then we don't need the `state === Complete || state === Completed` which is pointless).

If someone still wants to work with the 0160 state values, we could add a property that returns the 0160 state derived from the 0023 state if the user would want that. I would like to be able to use the DIDExchange and Connection protocol with as minimal API differences as possible. We call it the connections module because it allows me to create connections, whether that is through the DIDExchange or Connection protocol is irrelevant most of the time.

So maybe we can look at a way where:
- We only store a single value
- We only use that value around the framework to reduce complexity of state checks
- We add convenience methods to get the state in 0160 notation

What do you think of this approach?

Reply:
Let me just reiterate if I understand what you suggest. You also have a related suggestion here https://github.com/hyperledger/aries-framework-javascript/pull/615/files#r796613625, so I want to make it clear and in one place:

* We use `DidExchange` roles and states.
* OOB methods sets initial values in the following way:
`createMessage` **Responder**, invitation-sent
`receiveMessage` **Requester**, invitation-received
* We transform old Connection values to DidExchange values. There are 2 options:
	1. We prepare migration script to transform old roles and state to the new ones. This will force any current user to update event listeners to new API.
	2. We transform old Connection protocol values to DidExchange values if user requires to do that. How do we do that?

Transition table from DidExchange to Connection will look like this:
* start -> **null**
No connection exists or is in progress

* invitation-received, invitation-sent -> **invited**
The invitation has been shared with the intended *invitee*(s), and they have not yet sent a *connection_request*.

* request-sent, request-received -> **requested**
A *connection_request* has been sent by the *invitee* to the *inviter* based on the information in the *invitation*.

* response-sent, response-received  -> **responded**
A *connection_response* has been sent by the *inviter* to the *invitee* based on the information in the *connection_request*.

* completed -> **complete**

* abandoned -> ??? (`null`?)

Reply:
Yes that's what I'm suggesting. Re the transformation to connection state I would suggest adding a new computed property that returns the 0160 state

So this:

```
agent.events.on<ConnectionStateChangedEvent>(ConnectionEventTypes.StateChanged, (event) => {
    event.payload.record.state == ConnectionState.Requested
})
```

becomes this:

```
agent.events.on<ConnectionStateChangedEvent>(ConnectionEventTypes.StateChanged, (event) => {
    event.payload.record.rfc160State == ConnectionState.Requested
})
```